flow-model-generator/modules/flow-model-generator-ls-extension/src/main/java/io/ballerina/flowmodelgenerator/extension/ExpressionEditorService.java

```diff
@@ -132,84 +123,47 @@ public CompletableFuture<ExpressionEditorTypeResponse> types(VisibleVariableType
-
-                // Get signature help from language server
-                CompletableFuture<SignatureHelp> completableFuture =
-                        langServer.getTextDocumentService().signatureHelp(params);
-                SignatureHelp signatureHelp = completableFuture.join();
-
-                // Restore original content
-                context.applyContent(oldTextDocument);
-
-                return signatureHelp;
-            } catch (Throwable e) {
-                return new SignatureHelp();
-            }
-        });
+        String fileUri = CommonUtils.getExprUri(request.filePath());
+        return Debouncer.getInstance().debounce(
+                new SignatureHelpRequest(
+                        workspaceManagerProxy.get(fileUri),
+                        Path.of(request.filePath()),
+                        request.context(),
+                        fileUri,
+                        request.signatureHelpContext(),
+                        langServer.getTextDocumentService()
+                )
+        );
     }
 
     @JsonRequest
     public CompletableFuture<Either<List<CompletionItem>, CompletionList>> completion(
             ExpressionEditorCompletionRequest request) {
-        return CompletableFuture.supplyAsync(() -> {
-            try {
-                // Load the original project
-                Path filePath = Path.of(request.filePath());
-                String fileUri = CommonUtils.getExprUri(request.filePath());
-
-                // Get the document
-                Optional<Document> document = workspaceManagerProxy.get(fileUri).document(filePath);
-                if (document.isEmpty()) {
-                    return Either.forLeft(List.of());
-                }
-                TextDocument oldTextDocument = document.get().textDocument();
-
-                // Generate completions using context
-                ExpressionEditorContext context = new ExpressionEditorContext(workspaceManagerProxy.get(fileUri),
-                        request.context(), filePath, document.get());
-                context.generateStatement();
-
-                // Generate the completion params
-                Position position = context.getCursorPosition();
-                TextDocumentIdentifier identifier = new TextDocumentIdentifier(fileUri);
-                CompletionParams params = new CompletionParams(identifier, position, request.completionContext());
-
-                // Get completions from language server
-                CompletableFuture<Either<List<CompletionItem>, CompletionList>> completableFuture =
-                        langServer.getTextDocumentService().completion(params);
-                Either<List<CompletionItem>, CompletionList> completions = completableFuture.join();
-
-                // Restore original content
-                context.applyContent(oldTextDocument);
+        String fileUri = CommonUtils.getExprUri(request.filePath());
+        CompletionRequest completionApi =
```

Comment:
Shouldn't this be
```suggestion
        CompletionRequest completionRequest =
```
?